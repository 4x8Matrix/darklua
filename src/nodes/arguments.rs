use crate::nodes::{
    Expression,
    StringExpression,
    TableExpression,
};

use std::mem;

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Arguments {
    Tuple(Vec<Expression>),
    String(StringExpression),
    Table(TableExpression),
}

impl Arguments {
    pub fn to_expressions(self) -> Vec<Expression> {
        match self {
            Self::Tuple(expressions) => expressions,
            Self::String(string) => vec![string.into()],
            Self::Table(table) => vec![table.into()],
        }
    }

    pub fn append_argument<T: Into<Expression>>(self, argument: T) -> Self {
        let mut expressions = self.to_expressions();
        expressions.push(argument.into());
        Arguments::Tuple(expressions)
    }

    pub fn push_argument<T: Into<Expression>>(&mut self, argument: T) -> &mut Self {
        match self {
            Self::Tuple(expressions) => {
                expressions.push(argument.into());
            }
            Self::Table(_) | Self::String(_) => {
                match mem::replace(self, Self::Tuple(Vec::new())) {
                    Self::Table(table) => {
                        self.push_argument(table);
                    }
                    Self::String(string) => {
                        self.push_argument(string);
                    }
                    Self::Tuple(_) => {
                        unreachable!()
                    }
                }
                self.push_argument(argument);
            }
        }
        self
    }
}

impl Default for Arguments {
    fn default() -> Self {
        Arguments::Tuple(Vec::new())
    }
}
