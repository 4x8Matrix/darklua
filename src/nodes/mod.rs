//! The collection of nodes used for the Lua abstract syntax tree.

mod arguments;
mod block;
mod expressions;
mod function_call;
mod identifier;
mod statements;
mod token;
mod variable;

pub use arguments::*;
pub use block::*;
pub use expressions::*;
pub use function_call::*;
pub use identifier::*;
pub use statements::*;
pub use token::*;
pub use variable::*;

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum AnyNodeRef<'a> {
    AnyBlock(&'a Block),
    AnyExpression(AnyExpressionRef<'a>),
    AnyStatement(AnyStatementRef<'a>),
}

impl<'a> From<&'a Block> for AnyNodeRef<'a> {
    fn from(block: &'a Block) -> Self {
        Self::AnyBlock(block)
    }
}

impl<'a> From<&'a Statement> for AnyNodeRef<'a> {
    fn from(statement: &'a Statement) -> Self {
        Self::AnyStatement(AnyStatementRef::from(statement))
    }
}

impl<'a> From<&'a LastStatement> for AnyNodeRef<'a> {
    fn from(statement: &'a LastStatement) -> Self {
        Self::AnyStatement(AnyStatementRef::from(statement))
    }
}

impl<'a> From<&'a Expression> for AnyNodeRef<'a> {
    fn from(expression: &'a Expression) -> Self {
        Self::AnyExpression(AnyExpressionRef::from(expression))
    }
}

impl<'a> From<&'a Prefix> for AnyNodeRef<'a> {
    fn from(prefix: &'a Prefix) -> Self {
        Self::AnyExpression(AnyExpressionRef::from(prefix))
    }
}

impl<'a> From<&'a Arguments> for AnyNodeRef<'a> {
    fn from(arguments: &'a Arguments) -> Self {
        Self::AnyExpression(AnyExpressionRef::from(arguments))
    }
}

impl<'a> From<&'a Variable> for AnyNodeRef<'a> {
    fn from(variable: &'a Variable) -> Self {
        Self::AnyExpression(AnyExpressionRef::from(variable))
    }
}

impl<'a> From<&'a TableEntry> for AnyNodeRef<'a> {
    fn from(entry: &'a TableEntry) -> Self {
        Self::AnyExpression(AnyExpressionRef::from(entry))
    }
}

impl<'a> From<AnyStatementRef<'a>> for AnyNodeRef<'a> {
    fn from(any_statement: AnyStatementRef<'a>) -> Self {
        Self::AnyStatement(any_statement)
    }
}

impl<'a> From<AnyExpressionRef<'a>> for AnyNodeRef<'a> {
    fn from(any_expression: AnyExpressionRef<'a>) -> Self {
        Self::AnyExpression(any_expression)
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum AnyStatementRef<'a> {
    Statement(&'a Statement),
    LastStatement(&'a LastStatement),
}

impl<'a> From<&'a Statement> for AnyStatementRef<'a> {
    fn from(statement: &'a Statement) -> Self {
        Self::Statement(statement)
    }
}

impl<'a> From<&'a LastStatement> for AnyStatementRef<'a> {
    fn from(statement: &'a LastStatement) -> Self {
        Self::LastStatement(statement)
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum AnyExpressionRef<'a> {
    Expression(&'a Expression),
    Prefix(&'a Prefix),
    Arguments(&'a Arguments),
    Variable(&'a Variable),
    TableEntry(&'a TableEntry),
}

impl<'a> From<&'a Expression> for AnyExpressionRef<'a> {
    fn from(expression: &'a Expression) -> Self {
        Self::Expression(expression)
    }
}

impl<'a> From<&'a Prefix> for AnyExpressionRef<'a> {
    fn from(prefix: &'a Prefix) -> Self {
        Self::Prefix(prefix)
    }
}

impl<'a> From<&'a Arguments> for AnyExpressionRef<'a> {
    fn from(arguments: &'a Arguments) -> Self {
        Self::Arguments(arguments)
    }
}

impl<'a> From<&'a Variable> for AnyExpressionRef<'a> {
    fn from(variable: &'a Variable) -> Self {
        Self::Variable(variable)
    }
}

impl<'a> From<&'a TableEntry> for AnyExpressionRef<'a> {
    fn from(entry: &'a TableEntry) -> Self {
        Self::TableEntry(entry)
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum AnyNodeRefMut<'a> {
    AnyBlock(&'a mut Block),
    AnyExpression(AnyExpressionRefMut<'a>),
    AnyStatement(AnyStatementRefMut<'a>),
}

impl<'a> From<&'a mut Block> for AnyNodeRefMut<'a> {
    fn from(block: &'a mut Block) -> Self {
        Self::AnyBlock(block)
    }
}

impl<'a> From<&'a mut Statement> for AnyNodeRefMut<'a> {
    fn from(statement: &'a mut Statement) -> Self {
        Self::AnyStatement(AnyStatementRefMut::from(statement))
    }
}

impl<'a> From<&'a mut LastStatement> for AnyNodeRefMut<'a> {
    fn from(statement: &'a mut LastStatement) -> Self {
        Self::AnyStatement(AnyStatementRefMut::from(statement))
    }
}

impl<'a> From<&'a mut Expression> for AnyNodeRefMut<'a> {
    fn from(expression: &'a mut Expression) -> Self {
        Self::AnyExpression(AnyExpressionRefMut::from(expression))
    }
}

impl<'a> From<AnyStatementRefMut<'a>> for AnyNodeRefMut<'a> {
    fn from(any_statement: AnyStatementRefMut<'a>) -> Self {
        Self::AnyStatement(any_statement)
    }
}

impl<'a> From<AnyExpressionRefMut<'a>> for AnyNodeRefMut<'a> {
    fn from(any_expression: AnyExpressionRefMut<'a>) -> Self {
        Self::AnyExpression(any_expression)
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum AnyStatementRefMut<'a> {
    Statement(&'a mut Statement),
    LastStatement(&'a mut LastStatement),
}

impl<'a> From<&'a mut Statement> for AnyStatementRefMut<'a> {
    fn from(statement: &'a mut Statement) -> Self {
        Self::Statement(statement)
    }
}

impl<'a> From<&'a mut LastStatement> for AnyStatementRefMut<'a> {
    fn from(statement: &'a mut LastStatement) -> Self {
        Self::LastStatement(statement)
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum AnyExpressionRefMut<'a> {
    Expression(&'a mut Expression),
    Prefix(&'a mut Prefix),
    Arguments(&'a mut Arguments),
    Variable(&'a mut Variable),
    TableEntry(&'a mut TableEntry),
}

impl<'a> From<&'a mut Expression> for AnyExpressionRefMut<'a> {
    fn from(expression: &'a mut Expression) -> Self {
        Self::Expression(expression)
    }
}

impl<'a> From<&'a mut Prefix> for AnyExpressionRefMut<'a> {
    fn from(prefix: &'a mut Prefix) -> Self {
        Self::Prefix(prefix)
    }
}

impl<'a> From<&'a mut Arguments> for AnyExpressionRefMut<'a> {
    fn from(arguments: &'a mut Arguments) -> Self {
        Self::Arguments(arguments)
    }
}

impl<'a> From<&'a mut Variable> for AnyExpressionRefMut<'a> {
    fn from(variable: &'a mut Variable) -> Self {
        Self::Variable(variable)
    }
}

impl<'a> From<&'a mut TableEntry> for AnyExpressionRefMut<'a> {
    fn from(entry: &'a mut TableEntry) -> Self {
        Self::TableEntry(entry)
    }
}
