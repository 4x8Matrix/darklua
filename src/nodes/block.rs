use std::cmp::Ordering;

use crate::nodes::{AnyStatementRef, LastStatement, ReturnStatement, Statement, Token};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct BlockTokens {
    pub semicolons: Vec<Option<Token>>,
    pub last_semicolon: Option<Token>,
}

impl BlockTokens {
    pub fn clear_comments(&mut self) {
        self.semicolons.iter_mut().for_each(|semicolon| {
            if let Some(semicolon) = semicolon {
                semicolon.clear_comments();
            }
        });
        if let Some(last_semicolon) = &mut self.last_semicolon {
            last_semicolon.clear_comments();
        }
    }

    pub fn clear_whitespaces(&mut self) {
        self.semicolons.iter_mut().for_each(|semicolon| {
            if let Some(semicolon) = semicolon {
                semicolon.clear_whitespaces();
            }
        });
        if let Some(last_semicolon) = &mut self.last_semicolon {
            last_semicolon.clear_whitespaces();
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Block {
    statements: Vec<Statement>,
    last_statement: Option<LastStatement>,
    tokens: Option<Box<BlockTokens>>,
}

impl Block {
    pub fn new(statements: Vec<Statement>, last_statement: Option<LastStatement>) -> Self {
        Self {
            statements,
            last_statement,
            tokens: None,
        }
    }

    pub fn with_tokens(mut self, tokens: BlockTokens) -> Self {
        self.tokens = Some(tokens.into());
        self
    }

    #[inline]
    pub fn set_tokens(&mut self, tokens: BlockTokens) {
        self.tokens = Some(tokens.into());
    }

    #[inline]
    pub fn get_tokens(&self) -> Option<&BlockTokens> {
        self.tokens.as_ref().map(|tokens| tokens.as_ref())
    }

    pub fn with_statement<T: Into<Statement>>(mut self, statement: T) -> Self {
        self.statements.push(statement.into());
        self
    }

    pub fn with_last_statement<T: Into<LastStatement>>(mut self, last_statement: T) -> Self {
        self.last_statement = Some(last_statement.into());
        self
    }

    #[inline]
    pub fn is_empty(&self) -> bool {
        self.last_statement.is_none() && self.statements.is_empty()
    }

    #[inline]
    pub fn total_len(&self) -> usize {
        self.statements.len() + if self.last_statement.is_some() { 1 } else { 0 }
    }

    #[inline]
    pub fn iter_statements(&self) -> impl Iterator<Item = &Statement> {
        self.statements.iter()
    }

    #[inline]
    pub fn get_last_statement(&self) -> Option<&LastStatement> {
        self.last_statement.as_ref()
    }

    pub fn filter_statements<F>(&mut self, mut f: F)
    where
        F: FnMut(&Statement) -> bool,
    {
        let mut i = 0;

        while i != self.statements.len() {
            if f(&self.statements[i]) {
                i += 1;
            } else {
                self.statements.remove(i);

                if let Some(tokens) = &mut self.tokens {
                    if i < tokens.semicolons.len() {
                        tokens.semicolons.remove(i);
                    }
                }
            }
        }
    }

    pub fn filter_mut_statements<F>(&mut self, mut f: F)
    where
        F: FnMut(&mut Statement) -> bool,
    {
        let mut i = 0;

        while i != self.statements.len() {
            if f(&mut self.statements[i]) {
                i += 1;
            } else {
                self.statements.remove(i);

                if let Some(tokens) = &mut self.tokens {
                    if i < tokens.semicolons.len() {
                        tokens.semicolons.remove(i);
                    }
                }
            }
        }
    }

    pub fn truncate(&mut self, length: usize) {
        self.statements.truncate(length);
        if let Some(tokens) = &mut self.tokens {
            tokens.semicolons.truncate(length);
        }
    }

    #[inline]
    pub fn iter_mut_statements(&mut self) -> impl Iterator<Item = &mut Statement> {
        self.statements.iter_mut()
    }

    pub fn take_statements(&mut self) -> Vec<Statement> {
        if let Some(tokens) = &mut self.tokens {
            tokens.semicolons.clear();
        }
        self.statements.drain(..).collect()
    }

    pub fn take_last_statement(&mut self) -> Option<LastStatement> {
        if let Some(tokens) = &mut self.tokens {
            tokens.last_semicolon.take();
        }
        self.last_statement.take()
    }

    pub fn set_statements(&mut self, statements: Vec<Statement>) {
        self.statements = statements;

        if let Some(tokens) = &mut self.tokens {
            tokens.semicolons.clear();
        }
    }

    #[inline]
    pub fn mutate_last_statement(&mut self) -> Option<&mut LastStatement> {
        self.last_statement.as_mut()
    }

    #[inline]
    pub fn replace_last_statement<S: Into<LastStatement>>(
        &mut self,
        statement: S,
    ) -> Option<LastStatement> {
        self.last_statement.replace(statement.into())
    }

    pub fn get_statement(&self, index: usize) -> Option<AnyStatementRef> {
        match index.cmp(&self.statements.len()) {
            Ordering::Less => self.statements.get(index).map(AnyStatementRef::from),
            Ordering::Equal => self.last_statement.as_ref().map(AnyStatementRef::from),
            Ordering::Greater => None,
        }
    }

    pub fn clear(&mut self) {
        self.statements.clear();
        self.last_statement.take();

        if let Some(tokens) = &mut self.tokens {
            tokens.semicolons.clear();
            tokens.last_semicolon = None;
        }
    }

    pub fn clear_comments(&mut self) {
        if let Some(tokens) = &mut self.tokens {
            tokens.clear_comments();
        }
    }

    pub fn clear_whitespaces(&mut self) {
        if let Some(tokens) = &mut self.tokens {
            tokens.clear_whitespaces();
        }
    }
}

impl Default for Block {
    fn default() -> Self {
        Self::new(Vec::new(), None)
    }
}

impl<IntoStatement: Into<Statement>> From<IntoStatement> for Block {
    fn from(statement: IntoStatement) -> Block {
        Block::new(vec![statement.into()], None)
    }
}

impl From<LastStatement> for Block {
    fn from(statement: LastStatement) -> Block {
        Block::new(Vec::new(), Some(statement))
    }
}

impl From<ReturnStatement> for Block {
    fn from(statement: ReturnStatement) -> Block {
        Block::new(Vec::new(), Some(statement.into()))
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::nodes::DoStatement;

    #[test]
    fn default_block_is_empty() {
        let block = Block::default();

        assert!(block.is_empty());
    }

    #[test]
    fn is_empty_is_true_when_block_has_no_statements_or_last_statement() {
        let block = Block::new(Vec::new(), None);

        assert!(block.is_empty());
    }

    #[test]
    fn is_empty_is_false_when_block_has_a_last_statement() {
        let block = Block::default().with_last_statement(LastStatement::new_break());

        assert!(!block.is_empty());
    }

    #[test]
    fn is_empty_is_false_when_block_a_statement() {
        let block = Block::default().with_statement(DoStatement::default());

        assert!(!block.is_empty());
    }

    #[test]
    fn clear_removes_statements() {
        let mut block = Block::default().with_statement(DoStatement::default());
        block.clear();

        assert!(block.is_empty());
    }

    #[test]
    fn clear_removes_last_statement() {
        let mut block = Block::default().with_last_statement(LastStatement::new_break());
        block.clear();

        assert!(block.is_empty());
    }

    #[test]
    fn clean_removes_semicolon_tokens() {
        let mut block = Block::default()
            .with_statement(DoStatement::default())
            .with_tokens(BlockTokens {
                semicolons: vec![Some(Token::from_content(";"))],
                last_semicolon: None,
            });
        block.clear();

        assert!(block.get_tokens().unwrap().semicolons.is_empty());
    }

    #[test]
    fn clean_removes_last_semicolon_token() {
        let mut block = Block::default()
            .with_last_statement(LastStatement::new_break())
            .with_tokens(BlockTokens {
                semicolons: Vec::new(),
                last_semicolon: Some(Token::from_content(";")),
            });
        block.clear();

        assert!(block.get_tokens().unwrap().last_semicolon.is_none());
    }

    #[test]
    fn set_statements_clear_semicolon_tokens() {
        let mut block = Block::default()
            .with_statement(DoStatement::default())
            .with_tokens(BlockTokens {
                semicolons: vec![Some(Token::from_content(";"))],
                last_semicolon: None,
            });
        block.set_statements(Vec::new());

        assert!(block.get_tokens().unwrap().semicolons.is_empty());
    }

    #[test]
    fn take_last_statement_clear_semicolon_token() {
        let mut block = Block::default()
            .with_last_statement(LastStatement::new_break())
            .with_tokens(BlockTokens {
                semicolons: Vec::new(),
                last_semicolon: Some(Token::from_content(";")),
            });

        assert_eq!(
            block.take_last_statement(),
            Some(LastStatement::new_break())
        );

        assert!(block.get_tokens().unwrap().last_semicolon.is_none());
    }

    #[test]
    fn filter_statements_does_not_panic_when_semicolons_do_not_match() {
        let mut block = Block::default()
            .with_statement(DoStatement::default())
            .with_statement(DoStatement::default())
            .with_tokens(BlockTokens {
                semicolons: vec![Some(Token::from_content(";"))],
                last_semicolon: None,
            });

        block.filter_statements(|_statement| false);

        pretty_assertions::assert_eq!(
            block,
            Block::default().with_tokens(BlockTokens {
                semicolons: Vec::new(),
                last_semicolon: None,
            })
        );
    }

    #[test]
    fn filter_mut_statements_does_not_panic_when_semicolons_do_not_match() {
        let mut block = Block::default()
            .with_statement(DoStatement::default())
            .with_statement(DoStatement::default())
            .with_tokens(BlockTokens {
                semicolons: vec![Some(Token::from_content(";"))],
                last_semicolon: None,
            });

        block.filter_mut_statements(|_statement| false);

        pretty_assertions::assert_eq!(
            block,
            Block::default().with_tokens(BlockTokens {
                semicolons: Vec::new(),
                last_semicolon: None,
            })
        );
    }

    #[test]
    fn get_statement_of_empty_block_is_none() {
        let block = Block::default();

        assert_eq!(block.get_statement(0), None);
        assert_eq!(block.get_statement(2), None);
    }

    #[test]
    fn get_statement_outside_of_boundary() {
        let block = Block::default().with_statement(DoStatement::default());

        assert_eq!(block.get_statement(1), None);
    }

    #[test]
    fn get_statement_outside_of_boundary_with_last_statement() {
        let block = Block::default()
            .with_statement(DoStatement::default())
            .with_last_statement(LastStatement::new_break());

        assert_eq!(block.get_statement(2), None);
    }

    #[test]
    fn get_statement_returns_first_statement() {
        let block = Block::default().with_statement(DoStatement::default());

        assert_eq!(
            block.get_statement(0).unwrap(),
            AnyStatementRef::from(&Statement::Do(DoStatement::default().into()))
        );
    }

    #[test]
    fn get_statement_returns_first_when_it_is_the_last_statement() {
        let block = Block::default().with_last_statement(ReturnStatement::default());

        assert_eq!(
            block.get_statement(0).unwrap(),
            AnyStatementRef::from(&LastStatement::Return(ReturnStatement::default()))
        );
    }
}
