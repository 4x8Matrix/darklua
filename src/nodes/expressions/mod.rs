mod binary;
mod field;
mod function;
mod index;
mod lux;
mod number;
mod prefix;
mod string;
mod table;
mod unary;

pub use binary::*;
pub use field::*;
pub use function::*;
pub use index::*;
pub use lux::*;
pub use number::*;
pub use prefix::*;
pub use string::*;
pub use table::*;
pub use unary::*;

use crate::nodes::FunctionCall;

use std::num::FpCategory;

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Expression {
    Binary(Box<BinaryExpression>),
    Call(Box<FunctionCall>),
    False,
    Field(Box<FieldExpression>),
    Function(FunctionExpression),
    Identifier(String),
    Index(Box<IndexExpression>),
    Nil,
    Number(NumberExpression),
    Parenthese(Box<Expression>),
    String(StringExpression),
    Table(TableExpression),
    True,
    Unary(Box<UnaryExpression>),
    VariableArguments,
    LUX(LUXExpression),
}

impl From<bool> for Expression {
    fn from(boolean: bool) -> Expression {
        if boolean { Expression::True } else { Expression::False }
    }
}

impl From<f64> for Expression {
    fn from(value: f64) -> Expression {
        match value.classify() {
            FpCategory::Nan => {
                BinaryExpression::new(
                    BinaryOperator::Slash,
                    DecimalNumber::new(0.0),
                    DecimalNumber::new(0.0),
                ).into()
            }
            FpCategory::Infinite => {
                BinaryExpression::new(
                    BinaryOperator::Slash,
                    Expression::from(if value.is_sign_positive() { 1.0 } else { -1.0 }),
                    DecimalNumber::new(0.0),
                ).into()
            }
            FpCategory::Zero => {
                DecimalNumber::new(0.0).into()
            }
            FpCategory::Subnormal | FpCategory::Normal => {
                if value < 0.0 {
                    UnaryExpression::new(
                        UnaryOperator::Minus,
                        Expression::from(value.abs()),
                    ).into()
                } else {
                    if value < 0.1 {
                        let exponent = value.log10().floor();
                        let new_value = value / 10_f64.powf(exponent);

                        DecimalNumber::new(new_value)
                            .with_exponent(exponent as i64, true)
                            .into()
                    } else if value > 999.0 && (value / 100.0).fract() == 0.0 {
                        let mut exponent = value.log10().floor();
                        let mut power = 10_f64.powf(exponent);

                        while exponent > 2.0 && (value / power).fract() != 0.0 {
                            exponent -= 1.0;
                            power /= 10.0;
                        }

                        DecimalNumber::new(value / power)
                            .with_exponent(exponent as i64, true)
                            .into()
                    } else {
                        DecimalNumber::new(value).into()
                    }
                }
            }
        }
    }
}

impl From<BinaryExpression> for Expression {
    fn from(binary: BinaryExpression) -> Expression {
        Expression::Binary(Box::new(binary))
    }
}

impl From<FunctionCall> for Expression {
    fn from(call: FunctionCall) -> Expression {
        Expression::Call(Box::new(call))
    }
}

impl From<FieldExpression> for Expression {
    fn from(field: FieldExpression) -> Expression {
        Expression::Field(Box::new(field))
    }
}

impl From<FunctionExpression> for Expression {
    fn from(function: FunctionExpression) -> Self {
        Expression::Function(function)
    }
}

impl From<IndexExpression> for Expression {
    fn from(index: IndexExpression) -> Self {
        Self::Index(Box::new(index))
    }
}

impl From<NumberExpression> for Expression {
    fn from(number: NumberExpression) -> Self {
        Self::Number(number)
    }
}

impl From<DecimalNumber> for Expression {
    fn from(number: DecimalNumber) -> Self {
        Self::Number(NumberExpression::Decimal(number))
    }
}

impl From<HexNumber> for Expression {
    fn from(number: HexNumber) -> Self {
        Self::Number(NumberExpression::Hex(number))
    }
}

impl From<Prefix> for Expression {
    fn from(prefix: Prefix) -> Self {
        match prefix {
            Prefix::Call(call) => Self::Call(Box::new(call)),
            Prefix::Field(field) => Self::Field(field),
            Prefix::Identifier(name) => Self::Identifier(name),
            Prefix::Index(index) => Self::Index(index),
            Prefix::Parenthese(expression) => Self::Parenthese(Box::new(expression)),
        }
    }
}

impl From<StringExpression> for Expression {
    fn from(string: StringExpression) -> Self {
        Self::String(string)
    }
}

impl From<TableExpression> for Expression {
    fn from(table: TableExpression) -> Self {
        Self::Table(table)
    }
}

impl From<UnaryExpression> for Expression {
    fn from(unary: UnaryExpression) -> Self {
        Self::Unary(Box::new(unary))
    }
}
