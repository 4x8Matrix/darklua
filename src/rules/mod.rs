//! A module that contains the different rules that mutates a Lua block.

mod context;
pub use context::Context;

// rules
mod call_parens;
mod compute_expression;
mod convert_lux_to_roact;
mod empty_do;
mod group_local;
mod inject_value;
mod method_def;
mod no_local_function;
mod rename_variables;
mod unused_if_branch;
mod unused_while;

pub use call_parens::*;
pub use compute_expression::*;
pub use convert_lux_to_roact::*;
pub use empty_do::*;
pub use group_local::*;
pub use inject_value::*;
pub use method_def::*;
pub use no_local_function::*;
pub use rename_variables::*;
pub use unused_if_branch::*;
pub use unused_while::*;

use crate::nodes::Block;

use serde::{Serialize, Serializer, Deserialize, Deserializer};
use serde::ser::SerializeMap;
use serde::de::{self, MapAccess, Visitor};
use std::fmt;
use std::str::FromStr;
use std::collections::HashMap;

/// In order to be able to weakly-type the properties of any rule, this enum makes it possible to
/// easily use serde to gather the value associated with a property.
#[derive(Debug, Serialize, Deserialize)]
#[serde(untagged)]
pub enum RulePropertyValue {
    Boolean(bool),
    String(String),
    Usize(usize),
    Float(f64),
    StringList(Vec<String>),
    None,
}

/// When implementing the configure method of the Rule trait, the method returns a result that uses
/// this error type.
#[derive(Debug, Clone)]
pub enum RuleConfigurationError {
    /// When a rule gets an unknown property. The string should be the unknown field name.
    UnexpectedProperty(String),
    /// When a rule has a required property. The string should be the field name.
    MissingProperty(String),
    /// When a property is associated with something else than an expected string. The string is
    /// the property name.
    StringExpected(String),
    /// When a property is associated with something else than an expected unsigned number. The
    /// string is the property name.
    UsizeExpected(String),
    /// When a property is associated with something else than an expected list of strings. The
    /// string is the property name.
    StringListExpected(String),
    /// When the value type is invalid. The string is the property name that was given the wrong
    /// value type.
    UnexpectedValueType(String),
}

impl fmt::Display for RuleConfigurationError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use RuleConfigurationError::*;

        match self {
            UnexpectedProperty(property) => write!(f, "unexpected field '{}'", property),
            MissingProperty(property) => write!(f, "missing required field '{}'", property),
            StringExpected(property) => write!(f, "string value expected for field '{}'", property),
            UsizeExpected(property) => write!(f, "unsigned integer expected for field '{}'", property),
            StringListExpected(property) => write!(f, "list of string expected for field '{}'", property),
            UnexpectedValueType(property) => write!(f, "unexpected type for field '{}'", property),
        }
    }
}

pub type RuleProperties = HashMap<String, RulePropertyValue>;

pub type RuleProcessResult = Result<(), Vec<String>>;

/// Defines an interface that will be used to mutate blocks and how to serialize and deserialize
/// the rule configuration.
pub trait Rule: RuleConfiguration {
    /// This method should mutate the given block to apply the rule.
    fn process(&self, block: &mut Block, context: &mut Context) -> RuleProcessResult;
}

pub trait RuleConfiguration {
    /// The rule deserializer will construct the default rule and then send the properties through
    /// this method to modify the behavior of the rule.
    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError>;
    /// This method should the unique name of the rule.
    fn get_name(&self) -> &'static str;
    /// For implementing the serialize trait on the Rule trait, this method should return all
    /// properties that differs from their default value.
    fn serialize_to_properties(&self) -> RuleProperties;
}

pub trait FlawlessRule {
    fn flawless_process(&self, block: &mut Block, context: &mut Context);
}

impl<T: FlawlessRule + RuleConfiguration> Rule for T {
    fn process(&self, block: &mut Block, context: &mut Context) -> RuleProcessResult {
        self.flawless_process(block, context);
        Ok(())
    }
}

/// A function to get the default rule stack for darklua. All the rules here must preserve all the
/// functionalities of the original code after being applied. They must guarantee that the
/// processed block will work as much as the original one.
pub fn get_default_rules() -> Vec<Box<dyn Rule>> {
    vec![
        Box::new(ComputeExpression::default()),
        Box::new(RemoveUnusedIfBranch::default()),
        Box::new(RemoveUnusedWhile::default()),
        Box::new(RemoveEmptyDo::default()),
        Box::new(RemoveMethodDefinition::default()),
        Box::new(ConvertLocalFunctionToAssign::default()),
        Box::new(GroupLocalAssignment::default()),
        Box::new(RenameVariables::default()),
        Box::new(RemoveFunctionCallParens::default()),
    ]
}

impl FromStr for Box<dyn Rule> {
    type Err = String;

    fn from_str(string: &str) -> Result<Self, Self::Err> {
        let rule: Box<dyn Rule> = match string {
            COMPUTE_EXPRESSIONS_RULE_NAME => Box::new(ComputeExpression::default()),
            CONVERT_LUX_TO_ROACT_CODE_RULE_NAME => Box::new(ConvertLUXToRoactCode::default()),
            CONVERT_LOCAL_FUNCTION_TO_ASSIGN_RULE_NAME => Box::new(ConvertLocalFunctionToAssign::default()),
            GROUP_LOCAL_ASSIGNMENT => Box::new(GroupLocalAssignment::default()),
            INJECT_GLOBAL_VALUE_RULE_NAME => Box::new(InjectGlobalValue::default()),
            REMOVE_EMPTY_DO_RULE_NAME => Box::new(RemoveEmptyDo::default()),
            REMOVE_FUNCTION_CALL_PARENS => Box::new(RemoveFunctionCallParens::default()),
            REMOVE_METHOD_DEFINITION_RULE_NAME => Box::new(RemoveMethodDefinition::default()),
            REMOVE_UNUSED_IF_BRANCH_RULE_NAME => Box::new(RemoveUnusedIfBranch::default()),
            REMOVE_UNUSED_WHILE_RULE_NAME => Box::new(RemoveUnusedWhile::default()),
            RENAME_VARIABLES_RULE_NAME => Box::new(RenameVariables::default()),
            _ => return Err(format!("invalid rule name: {}", string)),
        };

        Ok(rule)
    }
}

impl Serialize for Box<dyn Rule> {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let properties = self.serialize_to_properties();
        let property_count = properties.len();
        let rule_name = self.get_name();

        if property_count == 0 {
            serializer.serialize_str(rule_name)

        } else {
            let mut map = serializer.serialize_map(Some(property_count + 1))?;

            map.serialize_entry("rule", rule_name)?;

            let mut ordered: Vec<(String, RulePropertyValue)> = properties.into_iter().collect();

            ordered.sort_by(|a, b| a.0.cmp(&b.0));

            for (key, value) in ordered {
                map.serialize_entry(&key, &value)?;
            }

            map.end()
        }
    }
}

impl<'de> Deserialize<'de> for Box<dyn Rule> {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Box<dyn Rule>, D::Error> {

        struct StringOrStruct;

        impl<'de> Visitor<'de> for StringOrStruct {
            type Value = Box<dyn Rule>;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("rule name or rule object")
            }

            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E> where E: de::Error {
                let mut rule: Self::Value = FromStr::from_str(value)
                    .map_err(de::Error::custom)?;

                rule.configure(RuleProperties::new())
                    .map_err(de::Error::custom)?;

                Ok(rule)
            }

            fn visit_map<M>(self, mut map: M) -> Result<Self::Value, M::Error> where M: MapAccess<'de> {
                let mut rule_name = None;
                let mut properties = HashMap::new();

                while let Some(key) = map.next_key::<String>()? {
                    match key.as_str() {
                        "rule" => if rule_name.is_none() {
                            rule_name.replace(map.next_value::<String>()?);
                        } else {
                            return Err(de::Error::duplicate_field("rule"))
                        }
                        property => {
                            let value = map.next_value::<RulePropertyValue>()?;

                            if properties.insert(property.to_owned(), value).is_some() {
                                return Err(de::Error::custom(format!("duplicate field {} in rule object", property)))
                            }
                        }
                    }
                }

                if let Some(rule_name) = rule_name {
                    let mut rule: Self::Value = FromStr::from_str(&rule_name)
                        .map_err(de::Error::custom)?;

                    rule.configure(properties).map_err(de::Error::custom)?;

                    Ok(rule)
                } else {
                    Err(de::Error::missing_field("rule"))
                }
            }
        }

        deserializer.deserialize_any(StringOrStruct)
    }
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;

    #[test]
    fn snapshot_default_rules() {
        let rules = get_default_rules();

        assert_json_snapshot!("default_rules", rules);
    }
}
