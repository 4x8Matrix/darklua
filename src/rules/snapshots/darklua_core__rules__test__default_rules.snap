---
source: src/rules/mod.rs
expression: rules

---
[
  "remove_spaces",
  "remove_comments",
  "compute_expression",
  "remove_unused_if_branch",
  "remove_unused_while",
  "remove_empty_do",
  "remove_method_definition",
  "convert_index_to_field",
  "rename_variables",
  "remove_function_call_parens"
]
